Stratis.Bitcoin.Features.RPC/Controllers/RPCController.cs

```diff
@@ -0,0 +1,118 @@
+using Microsoft.Extensions.Logging;
+using NBitcoin.RPC;
+using Stratis.Bitcoin.Builder.Feature;
+using Stratis.Bitcoin.Utilities.JsonErrors;
+
+namespace Stratis.Bitcoin.Features.RPC.Controllers
+{
+    /// <summary>
+    /// Controller providing API operations on the RPC feature.
+    /// </summary>
+    [Route("api/[controller]")]
+    public class RPCController : Controller
+    {
+        /// <summary>Instance logger.</summary>
+        private readonly ILogger logger;
+
+        /// <summary>Full Node.</summary>
+        private readonly IFullNode fullNode;
+
+        /// <summary>Full Node.</summary>
+        private readonly RpcSettings rpcSettings;
+
+        /// <summary>MethodInfo cache.</summary>
+        private readonly Dictionary<string, MethodInfo> methodInfoCache;
+
+        /// Initializes a new instance of the object.
+        /// </summary>
+        /// <param name="loggerFactory">Factory to be used to create logger for the node.</param>
+        public RPCController(IFullNode fullNode, ILoggerFactory loggerFactory, RpcSettings rpcSettings)
+        {
+            this.fullNode = fullNode;
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+            this.rpcSettings = rpcSettings;
+            this.methodInfoCache = new Dictionary<string, MethodInfo>();
+        }
+
+        /// <summary>
+        /// Finds an RPC method by name by looking in all features assemblies of the full node.
+        /// </summary>
+        /// <param name="name">The name of the RPC method to find.</param>
+        /// <returns>The 'MethodInfo' associated with the RPC method or null if the method is not found.</returns>
+        private MethodInfo GetRPCMethod(string name)
+        {
+            if (this.methodInfoCache.TryGetValue(name, out MethodInfo methodInfo))
+                return methodInfo;
+
+            foreach (Assembly asm in this.fullNode.Services.Features.OfType<FullNodeFeature>().Select(x => x.GetType().GetTypeInfo().Assembly).Distinct())
+            {
+                methodInfo = asm.GetTypes()
+                    .Where(type => typeof(Controller).IsAssignableFrom(type))
+                    .SelectMany(type => type.GetMethods())
+                    .Where(method => method.IsPublic && !method.IsDefined(typeof(NonActionAttribute)) && method.GetCustomAttribute<ActionNameAttribute>()?.Name == name).FirstOrDefault();
+
+                if (methodInfo != null)
+                    break;
+            }
+
+            this.methodInfoCache[name] = methodInfo;
+            return methodInfo;
+        }
+
+        /// <summary>
+        /// Call an RPC method by name.
+        /// </summary>
+        /// <returns>A JSON result that varies depending on the RPC method.</returns>
+        [Route("callbyname")]
+        [HttpGet]
+        public IActionResult CallByName([FromQuery]string methodName)
+        {
+            try
+            {
+                MethodInfo methodInfo = GetRPCMethod(methodName);
```

Comment:
missing this.

Reply:
Fixed

Reply:
@jeremy,  I agree that the "listmethods" call would be useful and have added it. It mimics the way that RPCRouteHandler discovers RPC methods so we should be good with only discovering the relevant methods. Since the constructor is a little too early to discover the methods in the preferred way I am caching all the methods the first time they are required.